turbine/src/cluster_nodes.rs

```diff
@@ -434,23 +434,27 @@ impl<T> ClusterNodesCache<T> {
         ttl: Duration,
     ) -> Self {
         Self {
-            cache: Mutex::new(LruCache::new(cap)),
+            cache: RwLock::new(LruCache::new(cap)),
             ttl,
         }
     }
 }
 
 impl<T: 'static> ClusterNodesCache<T> {
     fn get_cache_entry(&self, epoch: Epoch) -> Arc<RwLock<CacheEntry<T>>> {
-        let mut cache = self.cache.lock().unwrap();
-        match cache.get(&epoch) {
-            Some(entry) => Arc::clone(entry),
-            None => {
-                let entry = Arc::default();
-                cache.put(epoch, Arc::clone(&entry));
-                entry
-            }
+        if let Some(entry) = self.cache.read().unwrap().get(&epoch) {
+            return Arc::clone(entry);
+        }
```

Comment:
any insight into if this `if` evaluates to true more frequently than not? 

Thinking out loud: I would assume so since it is an LRU cache. In which case this trade off seems fair. Great performance gain in common case, slight performance degradation in much less common case.

Reply:
yeah, 99%+ this `if` is true, because the shreds you receive are either from current epoch or the next epoch.